### PR TITLE
New native FormInput component

### PR DESCRIPTION
**Pull request by buoyad**
This adds a native-only FormInput component to common adapters. Also adds provisions for stories for components that are native only.

image

r? @keybase/react-hackers

### CODE REVIEW

File: shared/common-adapters/form-input.native.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+              }}
+            />
+          )}
+        <NativeTextInput
+          autoCorrect={this.props.autoCorrect}
+          autoFocus={this.props.autoFocus}
+          autoGrow={!!this.props.multiline && !!this.props.maxHeight}
+          value={this.state.value}
+          onChangeText={this._onChangeText}
+          placeholder={this.state.focused ? '' : this.props.label}
```
Comment (buoyad) [inline shared/common-adapters/form-input.native.js]: There's a bug here in multiline inputs where if the input is touched but unfocused when empty, the placeholder can disappear. I spent some time working on it but couldn't make any progress

File: shared/common-adapters/form-input.native.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import {NativeTextInput} from './native-wrappers.native'
+import {globalColors, globalMargins, globalStyles} from '../styles'
+import {getStyle} from './text'
+
+type State = {
+  value: string,
+  focused: boolean,
+}
+
+export const backgroundModeToColor = {
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: this should likely go into styles/index

File: shared/common-adapters/form-input.native.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
+// @flow
+import * as React from 'react'
+import {Box, Text} from '.'
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: we shouldn't import from . inside common adapters. creates import loops and in general we've been avoiding it. just import ./box and ./text

File: shared/common-adapters/form-input.native.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+class FormInput extends React.Component<Props, State> {
+  state: State
+  _input: any
+
+  constructor(props: Props) {
+    super(props)
+
+    this.state = {
+      value: this.props.value || '',
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: i think you can use defaultProps for this so value is always a string
then you can just remove this constructor and make state as an initialized member property

File: shared/common-adapters/form-input.native.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      focused: false,
+    }
+  }
+
+  getValue = (): string => {
+    return this.state.value || ''
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: this is already always a string

File: shared/common-adapters/form-input.native.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  setValue = (value: string) => {
+    this.setState({value: value || ''})
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: value is already always a string

File: shared/common-adapters/form-input.native.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  blur = () => {
+    this._input && this._input.blur()
+  }
+
+  select = () => {
+    this._input && this._input.select()
+  }
+
+  _onChangeText = (text: string) => {
+    this.setState({value: text || ''})
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: text is always a string

File: shared/common-adapters/form-input.native.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+    this.props.onChangeText && this.props.onChangeText(text || '')
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: here too

File: shared/common-adapters/form-input.native.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  _onKeyPress = (e: SyntheticKeyboardEvent<>) => {
+    if (this.props.onEnterKeyDown && e.key === 'Enter') {
+      this.props.onEnterKeyDown()
+    }
+  }
+
+  render() {
+    let backgroundColor = this.props.backgroundMode
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: const

File: shared/common-adapters/form-input.native.js [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+      ? backgroundModeToColor[this.props.backgroundMode]
+      : globalColors.white
+    return (
+      <Box
+        style={{
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: we should chat about not having dynamic styles like this on common components if possible. else it'll always render thrash even if its really not changing. with RN we can use the StyleSheet and make individual styles and you can use the style={[styles.common, this.maxHeight && styles.maxHeight, this.props.containerStyle]} etc

File: shared/common-adapters/form-input.native.js [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+          borderColor: globalColors.black_10,
+          minHeight: this.props.multiline ? 128 : 64,
+          ...(this.props.maxHeight ? {maxHeight: this.props.maxHeight} : {}),
+          ...this.props.containerStyle,
+        }}
+      >
+        {(!!this.state.value || this.state.focused) && (
+          <Text
+            type="BodySmallSemibold"
+            style={{
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: should be pulled out into a const

File: shared/common-adapters/form-input.native.js [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+            }}
+          >
+            {this.props.label}
+          </Text>
+        )}
+        {this.props.label &&
+          this.props.multiline && (
+            <Box
+              style={{
+                position: 'absolute',
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.native.js]: same here

File: shared/common-adapters/form-input.js.flow [hunk 13]
```diff
@@ -0,0 +1,27 @@
+// @flow
+import {Component} from 'react'
```
Comment (chrisnojima) [inline shared/common-adapters/form-input.js.flow]: since this is mobile only lets kill this file

**Pull request merged**